situp-core/src/main/java/com/amazon/situp/parser/PipelineConfigurationValidator.java

```diff
@@ -25,12 +36,133 @@ public static void validate(final PipelineConfiguration pipelineConfiguration) {
+            validate(pipelineConfiguration);
+            touchedPipelineSet.add(pipeline);
+            //if validation is successful, then there is definitely sink
+            final List<PluginSetting> connectedPipelinesSettings = pipelineConfiguration.getSink().getPluginSettings();
+            //Recursively check connected pipelines
+            for (PluginSetting pluginSetting : connectedPipelinesSettings) {
+                //Further process only if the sink is of pipeline type
+                if (pluginSetting.getName().equals(PIPELINE_TYPE)) {
+                    final String connectedPipelineName = (String) pluginSetting.getAttributeFromSettings(PIPELINE_ATTRIBUTE_NAME);
+                    if (connectedPipelineName == null || "".equals(connectedPipelineName.trim())) {
+                        throw new RuntimeException(format("name is a required attribute for sink pipeline plugin, " +
+                                "check pipeline: %s", pipeline));
+                    }
+                    validateSourceMapping(pipeline, connectedPipelineName, pipelineConfigurationMap);
+                    visitAndValidate(connectedPipelineName, pipelineConfigurationMap, touchedPipelineSet,
+                            visitedAndProcessedPipelineSet, sortedPipelineNames);
+                }
+            }
+            visitedAndProcessedPipelineSet.add(pipeline);
+            touchedPipelineSet.remove(pipeline);
+            sortedPipelineNames.add(pipeline);
+        }
+    }
+
+    /**
+     * This method validates if pipeline's source is correctly configured to reflect the sink of its parent i.e.
+     * if p2 is defined as sink for p1, source of p2 should be defined as p1.
+     *
+     * @param sourcePipeline           name of the expected source pipeline
+     * @param currentPipeline          name of the current pipeline that is being validated
+     * @param pipelineConfigurationMap pipeline name to pipeline configuration map
+     */
+    private static void validateSourceMapping(
+            final String sourcePipeline,
+            final String currentPipeline,
+            final Map<String, PipelineConfiguration> pipelineConfigurationMap) {
+        if (!pipelineConfigurationMap.containsKey(currentPipeline)) {
+            throw new RuntimeException(format("Invalid configuration, no pipeline is defined with name %s", currentPipeline));
+        }
+        final PipelineConfiguration pipelineConfiguration = pipelineConfigurationMap.get(currentPipeline);
+        //Current deserialization has empty source if it is not defined - so no NPE
+        final List<PluginSetting> sourcePluginSettings = pipelineConfiguration.getSource().getPluginSettings();
+        if (sourcePluginSettings.isEmpty() || sourcePluginSettings.get(0) == null ||
+                !PIPELINE_TYPE.equals(sourcePluginSettings.get(0).getName()) ||
+                !sourcePipeline.equals(sourcePluginSettings.get(0).getAttributeFromSettings(PIPELINE_ATTRIBUTE_NAME))) {
```

Comment:
I tried to avoid but had to move with it as this is happening before validation, if its after validation we will be sure that there is only one Source. I will try to update this logic